File: clang/test/CodeGenObjC/assign.m
// RUN: %clang_cc1 -triple x86_64 -emit-llvm -o - %s | FileCheck %s

struct s0 {
  int x;
};

@interface C0
@property int x0;
@property _Complex int x1;
@property struct s0 x2;
@end

// Check that we get exactly the message sends we expect, and no more.
//
// CHECK: define void @f0
void f0(C0 *a) {
// CHECK: objc_msgSend
  int l0 = (a.x0 = 1);

// CHECK: objc_msgSend
  _Complex int l1 = (a.x1 = 1);

// CHECK: objc_msgSend
  struct s0 l2 = (a.x2 = (struct s0) { 1 });

// CHECK: objc_msgSend
// CHECK: objc_msgSend
  int l3 = (a.x0 += 1);

// CHECK: objc_msgSend
// CHECK: objc_msgSend
  _Complex int l4 = (a.x1 += 1);

// CHECK-NOT: objc_msgSend
// CHECK: }
}

File: clang/test/CodeGen/assign.c
// RUN: %clang_cc1 -triple x86_64 -emit-llvm -o - %s | FileCheck %s

// Check that we don't generate unnecessary reloads.
//
// CHECK-LABEL: define void @f0()
// CHECK:      [[x_0:%.*]] = alloca i32, align 4
// CHECK-NEXT: [[y_0:%.*]] = alloca i32, align 4
// CHECK-NEXT: store i32 1, i32* [[x_0]]
// CHECK-NEXT: store i32 1, i32* [[x_0]]
// CHECK-NEXT: store i32 1, i32* [[y_0]]
// CHECK: }
void f0() {
  int x, y;
  x = 1;
  y = (x = 1);
}

// This used to test that we generate reloads for volatile access,
// but that does not appear to be correct behavior for C.
//
// CHECK-LABEL: define void @f1()
// CHECK:      [[x_1:%.*]] = alloca i32, align 4
// CHECK-NEXT: [[y_1:%.*]] = alloca i32, align 4
// CHECK-NEXT: store volatile i32 1, i32* [[x_1]]
// CHECK-NEXT: store volatile i32 1, i32* [[x_1]]
// CHECK-NEXT: store volatile i32 1, i32* [[y_1]]
// CHECK: }
void f1() {
  volatile int x, y;
  x = 1;
  y = (x = 1);
}
